### DISCUSSION TITLE
Refactor of the main user interface

### DISCUSSION THREAD

**Discussion by rlouf** [outlines-dev/outlines#1220] [General]
The current design of the library is not flexible enough:

1. A new output type requires a new function in the `outlines.generate` module;
2. Users need to remember the names of these different functions;
3. It is difficult to pass model-specific initialization and inference parameters.

## New user interface

We need to make the interface of the library simpler and more flexible. I propose the following design, in pseudo-code:

```python
from outlines import models


model = models.provider("name")
result = model(prompt, OutputType)
```

Users thus need only be concerned about the output type, be it a Python type, a Pydantic model, etc. without having to learn new functions. This implicitly re-centers Outlines around the definition of output types.

## Extra parameters

Any other value passed to `models.provider` is passed directly to the initialization function in the corresponding library:

```python
model = models.provider("name", *init_args, **init_kwargs)
```

Same for other values passed to the `__call__` method of the model:

```python
result = model(prompt, OutputType, *inference_args, **inference_kwargs)
```

This will give users more flexibility. For instance this would solve #1199, and would allow users to use a wider variety of sampling algorithms that those described in `samplers.py`. It would also simplify the code as we will not be trying to normalize the parameters anymore. See [here](https://github.com/dottxt-ai/outlines/blob/a2fa1e0f44e80dd287c215fffde342578e67b1d1/outlines/generate/api.py#L412), [here](https://github.com/dottxt-ai/outlines/blob/a2fa1e0f44e80dd287c215fffde342578e67b1d1/outlines/models/llamacpp.py#L110) or [here](https://github.com/dottxt-ai/outlines/blob/a2fa1e0f44e80dd287c215fffde342578e67b1d1/outlines/models/vllm.py#L89) for example.

Outlines will become a *thin wrapper* around the libraries to augment them with a friendly interface to do structured generation.

## Async execution

Asynchronous execution is necessary for agentic workflows, among other things. We should thus support async calls whenever possible:

- OpenAI
- Gemini
- vLLM. This will require us to wrap `AsyncLLMEngine`.

## Streaming

We should also offer the possibility to stream tokens, although I am not quite sure how that would work with types such as Pydantic models. A common way to do this is to pass `streaming=True` to the generation function:

```python
results = model(prompt, output_type, stream=True)
```

Although I am not a big fan of this and would prefer a new method such as:

```python
results = model.stream(prompt, output_type)
```

## Multi-modal models

Multi-modal models are different form text-2-text model in that they accept multiple modalities as an input. I thus believe they can simply be handled by defining specific input types:

```python
from outlines.inputs import Vision


model("prompt", output_type)  # Text to `output_type`
model(Vision("prompt", image), output_type)  # Text & Image to `output_type`
```

In this case, however, if `image` is of type `PIL.Image` we may be able to simply pass a tuple as an input:

```python
model(("prompt", image), output_type)
```

In any case, this should be handled by looking at the types of the inputs.

## Reviewers

@torymur, @lapp0

**Comment by torymur** [comment]
It seems to be worth it to take it to the full Builder Pattern, which will provide:
- flexibility: it would be easy to adjust it in any way
- clarity: separates all these concerns perfectly

Imagine this interface:
```python
from outlines import models, OutputType

model = models.provider("name", *init_args, **init_kwargs)
result = model \
    .inference_settings(*args, **kwargs) \ # optional
    .txt_prompt() | .visual_prompt() | .audio_prompt() \ # any or all
    .output_type(OutputType) \
    .stream() or .load() # final call, which is hidden build() + output approach
```

### Where:

#### 1. `.load()`
As in `json`/`pickle`, as one of the explicit alternative names to streaming, which is to say "buffering/batching", but in more pythonic way.
There are other names too, for example:
 - `fetch` as in `requests`:
```results = model.fetch(prompt, output_type)```

- `extract` as in `lxml`/`BeautifulSoup`: 
```results = model.extract(prompt, output_type)```

Or maybe something else. 

#### 2. Unsupported prompt types

Since it's difficult to determine what particular model can and cannot do in terms of types of prompts (txt, visual, video, audio, etc.), we could provide all types and catch the exceptions with nice consistent message of unsupported prompt type.

#### 3. Chaining prompts

Builder Pattern will also allow easily chain different prompt types, which might be quite useful.

**Reply by just-cameron** [reply]
> ```python
> from outlines import models, OutputType
> 
> model = models.provider("name", *init_args, **init_kwargs)
> result = model \
>     .inference_settings(*args, **kwargs) \ # optional
>     .txt_prompt() | .visual_prompt() | .audio_prompt() \ # any or all
>     .output_type(OutputType) \
>     .stream() or .load() # final call, which is hidden build() + output approach
> ```

I like the builder approach here, particularly since model providers are starting to add more custom token fields (reflection, tools available, etc) and we could flexibly extend this to handle common special token types.

**Reply by rlouf** [reply]
I think a better pattern to manage token types is to use an object that represents the prompt, and translating it to a prompt string before passing it to the model. This is what the [prompt](https://github.com/dottxt-ai/prompts) library is supposed to be doing.

**Reply by torymur** [reply]
I agree, that custom token fields seems to be a natural part of .txt input type, also I would imagine high inconsistency and great variability, since currently it's quite experimental, maybe it'll transform later into something different, let's see. And indeed `prompt` lib might shine even more considering these custom token fields, maybe we can extend examples with these too.

I suggested the pattern, even though it looks different to what we have currently, mainly to be **very** explicit of what interface provides and be highly flexible, but visually stable later.

For example, I agree with @rlouf that streaming approach without flags is better, but if we'll have `.stream`, then there is a necessity to being explicit of how to make the opposite call without guessing. Or this situation with different prompt types, I think having them defined separately will help distinguish, pass correct arguments and later easily debug, when model doesn't support visual prompts for instance and fails. So that, user experience will be smoother. Even though in python being implicit it's quite common as a pattern and people will likely guess correctly that a call without a `.stream` is the opposite one, I think it might worth to break out of it entirely to maximize predictability of UI.

**Reply by rlouf** [reply]
There is one user interface issue to note with your proposal: For long building expressions user will need to use `\` to break lines, or write the whole expression between parentheses. Not ideal.

But let's forget about this for a second and address the design itself. My problem with it boils down to: what does an object to which we haven't applied `stream` or `load` represent? Not really a result, which is a string. Not really a model conditioned on a prompt and output type, since we're mixing inference-time considerations when using `inference_settings`.

There is a third option that we haven't considered yet, which might solve the semantics issue. We could define `generate` and `stream` functions with the following pseudo-signatures:

```python
generate(
    model,
    input,
    output_type,
    *inference_args,
    **inference_kwargs
)
```

And

```python
stream(
    model,
    input,
    output_type,
    *inference_args,
    **inference_kwargs
)
```

It is very close to the current design, although more flexible since the output type becomes an argument. Like in the first design I introduced in the original post, we discriminate between the different kinds of inputs based on their type. We also avoid this awkward `stream` method.

Under the hood this will be implemented by calling `generate` and `stream` methods of the object class. I don't see a way around this as long as the sequence generation process is handled by the upstream libraries. Having said that, there is no reason to bubble up this constraint all the way up to the users and have them manipulate code with poor semantics.

**Reply by rlouf** [reply]
One piece of context I didn't mention: why do we currently generate text in two steps by first building a generator and then calling it?

```python
from outlines import generate

generator = generate.json(model, OutputModel)
result = generator("prompt")
```

The reason is that building the index is expensive, but it only needs to be computed once for every `(model, OutputType)`. We wanted to nudge the user so they don't repeatedly build the index. I see two solutions here:

1. Cache the index so the user does not have to be aware of this. The risk is to confuse the user: why are second runs faster than the first ones? Having said that, we already cache the index to disk.
2. Still build `SequenceGenerator` and `SequenceStreamer` objects with:

```python
generate = generator(model, output_type)
result = generate(prompt, *inference_args, **inference_kwargs)

stream = streamer(model, output_type)
result = stream(prompt, *inference_args, **inference_kwargs)

type(generate)
# SequenceGenerator

type(stream)
# SequenceStreamer
``` 

This has the further advantage of being only a minor change from the current design, while still simplifying it and the internals substantially.

We could also build a `generator` or `streamer` from an `Index` object instead of an output type.

**Reply by torymur** [reply]
> The reason is that building the index is expensive, but it only needs to be computed once for every (model, OutputType). We wanted to nudge the user so they don't repeatedly build the index.

Actually, this seems to be a key point from the interface perspective, since it's about efficient usability too.

> Cache the index so the user does not have to be aware of this. The risk is to confuse the user: why are second runs faster than the first ones? Having said that, we already cache the index to disk.

Looks like caching could be considered as a separate feature. And user might decide to disable cache, but even in that case interface needs to allow efficient calls with different prompts as many times as needed without costly index building call each time, yeah. Which hints to having something like:

```
from outlines import generator

generator = generator(model, output_type)
stream_result = generator.stream("prompt")
collect_result = generator.collect("another prompt")
```

Even if technically user might call it in one line, interface visually hints that these are intentionally separate steps:
```
from outlines import generator

stream_result = generator(model, output_type).stream("prompt")
collect_result = generator(model, output_type).collect("another prompt")
```

I did "misused" the `generator` from your example, which is actually perfect for both, since generator can be streamed or all at once collected (consumed).

> My problem with it boils down to: what does an object to which we haven't applied stream or load represent? Not really a result, which is a string. Not really a model conditioned on a prompt and output type, since we're mixing inference-time considerations when using inference_settings.

Just to answer this one: since we're essentially building the output, then maybe a Draft (or even Index?).
Although in the context of examples above it seems to be an Output Generator.

> We could also build a generator or streamer from an Index object instead of an output type.

How would it look like?

**Reply by rlouf** [reply]
> Looks like caching could be considered as a separate feature. And user might decide to disable cache, but even in that case interface needs to allow efficient calls with different prompts as many times as needed without costly index building call each time, yeah. 

Agreed.

> Which hints to having something like:

```python
from outlines import generator

generate = generator(model, output_type)
stream_result = generate.stream("prompt")
collect_result = generate.collect("another prompt")
```

Your design has a lot of merit, and I am wondering if we can push it a little further. What if we made `generator` an actual Python generator:

```python
generate = generator(model, output_type)
tokens = generate("prompt")

for token in tokens:
    print(token)
```

And we can consume the `Generator` object like you would any Python generator:

```python
result = consume(tokens)
```

This feel more "pythonic". The downside with this design, however, is that it requires a big shift in users' mental model who are used to get the full generation by default (Note that vLLM's `AsyncLLMEngine` [streams by default](https://docs.vllm.ai/en/latest/dev/engine/async_llm_engine.html))

> How would it look like?

If we have a pre-compiled `Index` object called `index` it would simply look like

```python
generate = generator(model, index)
```

And we'd have a separate function to compile an index based on a model and output type. There is a risk that people will use an index that was compiled for a given vocabulary with a model that has a different vocabulary, but we can probably do something about it by storing the model name along the index.

**Reply by torymur** [reply]
> What if we made generator an actual Python generator

It could be that too, if `generator(model, output_type)("prompt")` returns a pure python generator, then it could be dealt with as any other generator - be iterated over or consumed at once in usual manner: `list(tokens)`

>  The downside with this design, however, is that it requires a big shift in users' mental model who are used to get the full generation by default

Yeah, by nature it's unidirectional from stream to full, we can only offer different interface: generator or builder.
Hopefully, people are used to forcing full generation by `list(...)`.

> There is a risk that people will use an index that was compiled for a given vocabulary with a model that has a different vocabulary, but we can probably do something about it by storing the model name along the index.

If index is being computed for every `(model, OutputType)`, then it eliminates this mismatch problem, no?
It would be like: `generator(index)`

**Reply by rlouf** [reply]
Using `list(tokens)` to get the full generation is inconvenient in this case as you would get a list of tokens when the user generally wants `"".join(list(tokens))`

**Reply by rlouf** [reply]
Alternatively, `generate("prompt")` can return an `itertools.accumulate` object built from an iterator that returns tokens one by one. The full generation is then obtained as:

```python
*_, result = generate("prompt")
```

**Reply by torymur** [reply]
> user generally wants "".join(list(tokens))

Ah, you're right! Well, if only string types will be in generator, then it's possible to use shorter version: `''.join(token)`

So options are:
1) `''.join(tokens)`
2) alias helper, but import line won't be as simple: `from outlines import generator, consume` (or some other name instead of `consume`, but still a separate name)
3) custom iterator with helpers: `generate("prompt").join()`

> generate("prompt") can return an `itertools.accumulate` object

This will return "unexpected" values, which are not new on every iter, but accumulated already, which then prevent usage in a loop. This argument might be nullified, considering the overall weirdness of streaming showed by @cpfiffer

**Comment by just-cameron** [comment]
> We should also offer the possibility to stream tokens, although I am not quite sure how that would work with types such as Pydantic models. A common way to do this is to pass `streaming=True` to the generation function:
> 
> ```python
> results = model(prompt, output_type, stream=True)
> ```
> 
> Although I am not a big fan of this and would prefer a new method such as:
> 
> ```python
> results = model.stream(prompt, output_type)
> ```

For pydantic models, I've seen BAML handle streaming by generating the entire key structure with empty fields, and then filling out values as you go. This is presumably quite resource intensive, but it did give a cohesive streaming generation experience that respected structure. 

Example:

Step 1:

```json
{
  "field1":"",
  "field2":null,
}
```

Step 2, first sample

```json
{
  "field1":"Hello ",
  "field2":null,
}
```

Step 3, completion of first value

```json
{
  "field1":"Hello World",
  "field2":null,
}
```

Step 4, complete `field2`:

```json
{
  "field1":"Hello ",
  "field2":15,
}
```

Streaming in structured gen is very strange though, so I'm not sure what the appropriate interface is.

The bonuses of implementing a "fill the skeleton" approach as above is that, in principle, you can  to run early validation, inline tool evaluations, etc.

**Reply by torymur** [reply]
Indeed, that's not what I would expect as a stream, I guess we will be simply relaying whatever we get.

**Reply by just-cameron** [reply]
More on this: https://pypi.org/project/pydantic-partial/

I believe most other projects that fill out the skeleton (like instructor) use Pydantic partials.

**Reply by rlouf** [reply]
This discussion is closely tied to discussions about the output types, but the interface choice is independent from these considerations. Let's not get lost in the weeds and discuss it later.
